hpy/debug/src/_debugmod.c

```diff
@@ -0,0 +1,212 @@
+                return HPy_NULL;
+            }
+            HPy_Close(uctx, u_item);
+        }
+        dh = dh->next;
+    }
+    return u_result;
+}
+
+/* ~~~~~~ DebugHandleType and DebugHandleObject ~~~~~~~~
+
+   This is the applevel view of a DebugHandle/DHPy.
+
+   Note that there are two different ways to expose DebugHandle to applevel:
+
+   1. make DebugHandle itself a Python object: this is simple but means that
+      you have to pay the PyObject_HEAD overhead (16 bytes) for all of them
+
+   2. make DebugHandle a plain C struct, and expose them through a
+      Python-level wrapper.
+
+   We choose to implement solution 2 because we expect to have many
+   DebugHandle around, but to expose only few of them to applevel, when you
+   call get_open_handles. This way, we save 16 bytes per DebugHandle.
+
+   This means that you can have different DebugHandleObjects wrapping the same
+   DebugHandle. To make it easier to compare them, they expose the .id
+   attribute, which is the address of the wrapped DebugHandle. Also,
+   DebugHandleObjects compare equal if their .id is equal.
+*/
+
+typedef struct {
+    HPyObject_HEAD
+    DebugHandle *handle;
+} DebugHandleObject;
+
+HPyDef_GET(DebugHandle_obj, "obj", DebugHandle_obj_get,
+           .doc="The object which the handle points to")
+static UHPy DebugHandle_obj_get(HPyContext uctx, UHPy self, void *closure)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    return HPy_Dup(uctx, dh->handle->uh);
+}
+
+HPyDef_GET(DebugHandle_id, "id", DebugHandle_id_get,
+           .doc="A numeric identifier representing the underlying universal handle")
+static UHPy DebugHandle_id_get(HPyContext uctx, UHPy self, void *closure)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    return HPyLong_FromSsize_t(uctx, (HPy_ssize_t)dh->handle);
+}
+
+HPyDef_SLOT(DebugHandle_cmp, DebugHandle_cmp_impl, HPy_tp_richcompare)
+static UHPy DebugHandle_cmp_impl(HPyContext uctx, UHPy self, UHPy o, HPy_RichCmpOp op)
+{
+    UHPy T = HPy_Type(uctx, self);
+    if (!HPy_TypeCheck(uctx, o, T))
+        return HPy_Dup(uctx, uctx->h_NotImplemented);
+    DebugHandleObject *dh_self = HPy_CAST(uctx, DebugHandleObject, self);
+    DebugHandleObject *dh_o = HPy_CAST(uctx, DebugHandleObject, o);
+
+    switch(op) {
+    case HPy_EQ:
+        return HPyBool_FromLong(uctx, dh_self->handle == dh_o->handle);
+    case HPy_NE:
+        return HPyBool_FromLong(uctx, dh_self->handle != dh_o->handle);
+    default:
+        return HPy_Dup(uctx, uctx->h_NotImplemented);
+    }
+}
+
+HPyDef_SLOT(DebugHandle_repr, DebugHandle_repr_impl, HPy_tp_repr)
+static UHPy DebugHandle_repr_impl(HPyContext uctx, UHPy self)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    UHPy uh_fmt = HPy_NULL;
+    UHPy uh_id = HPy_NULL;
+    UHPy uh_args = HPy_NULL;
+    UHPy uh_result = HPy_NULL;
+
+    uh_fmt = HPyUnicode_FromString(uctx, "<DebugHandle 0x%x for %r>");
```

Comment:
We could do with `PyUnicode_FromFormat` here.

Reply:
good point, although we don't have `HPyUnicode_FromFormat` here. And we don't want to use the `Py_*` version because this is supposed to be a pure-hpy module, remember it will be used also by pypy and hopefully graalpython.
I addedd an XXX in e2b548a